tidy3d/components/data/monitor_data.py

```diff
@@ -424,6 +424,194 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
+            data_template = self.sel_freq_index(freq_index=f0_ind)
+            norm_template = np.sqrt(np.abs(data_template.dot(data_template)))
+            norm_template = norm_template.data.ravel()
+            # March to lower/higher frequencies
+            for freq_id in range(f0_ind + step, last_ind, step):
+                # Get next frequency to sort
+                data_to_sort = self.sel_freq_index(freq_index=freq_id)
+                norm_to_sort = np.sqrt(np.abs(data_to_sort.dot(data_to_sort)))
+                norm_to_sort = norm_to_sort.data.ravel()
+                # Compute "sorting w.r.t. to neighbor" and overlap values
+                sorting_one_mode, overlap_one_mode = self._find_ordering_one_freq(
+                    data_template, norm_template, data_to_sort, norm_to_sort, overlap_thresh
+                )
+                # Transform "sorting w.r.t. neighbor" to "sorting w.r.t. to f0_ind"
+                sorting[freq_id, :] = sorting_one_mode[sorting[freq_id - step, :]]
+                overlap[freq_id, :] = overlap_one_mode[sorting[freq_id - step, :]]
+                # Check for discontinuities and show warning if any
+                for mode_ind in list(np.nonzero(overlap[freq_id, :] < overlap_thresh)[0]):
+                    log.warning(
+                        f"WARNING: Mode '{mode_ind}' appears to undergo a discontinuous change "
+                        f"between frequencies '{self.monitor.freqs[freq_id]}' "
+                        f"and '{self.monitor.freqs[freq_id - step]}' "
+                        f"(overlap: '{overlap[freq_id, mode_ind]:.2f}')."
+                    )
+                # Reassign for the next iteration
+                data_template = data_to_sort
+                norm_template = norm_to_sort
+
+        # Rearrange modes using computed sorting values
+        mode_data_sorted = self._reorder_modes(
+            sorting=sorting,
+            gauge_indices=gauge_indices,
+            track=track,
+        )
+
+        return mode_data_sorted
+
+    # pylint:disable=too-many-arguments
+    def _find_ordering_one_freq(
+        self,
+        data_template: ModeSolverData,
+        norm_template: Numpy,
+        data_to_sort: ModeSolverData,
+        norm_to_sort: Numpy,
+        overlap_thresh: float = 0.75,
+    ) -> Tuple[Numpy, Numpy]:
+        """Find new ordering of modes in data_to_sort based on their similarity to data_template."""
+
+        num_modes = data_template.n_complex.sizes["mode_index"]
+
+        # Check whether modes already match
+        current_overlaps = np.abs(data_template.dot(data_to_sort))
+        current_overlaps = current_overlaps.data.ravel() / norm_template / norm_to_sort
+
+        if np.all(current_overlaps > overlap_thresh):
+            return np.arange(num_modes), current_overlaps
+
+        # Compute an overlap matrix, dot products of each template mode with each mode to sort
```

Comment:
One thing I am wondering is if we should do this for all modes or only for the subset of modes with `current_overlap < overlap_thresh` and maybe setting a higher `overlap_thresh` by default. I understand that it's probably best to compute the full matrix but I'm just a bit worried about performance again in the case of a large number of modes in which case also it's likely that at least a few (e.g. the spurious ones) may have a low overlap.

Reply:
Sounds like a good idea, I think that should work. Can it happen that some two modes become degenerate at some frequencies? Maybe only in such cases it won't work perfectly

Reply:
Yeah it can happen. Is this hard to handle either way?